cmd/tendermint/commands/rollback.go

```diff
@@ -0,0 +1,171 @@
+
+	// Get State
+	stateDB, err := node.DefaultDBProvider(&node.DBContext{ID: "state", Config: config})
+	if err != nil {
+		return err
+	}
+
+	genDoc, err := loadGenesisDoc(stateDB)
+	if err != nil {
+		genDoc, err = node.DefaultGenesisDocProviderFunc(config)()
+		if err != nil {
+			return err
+		}
+		// save genesis doc to prevent a certain class of user errors (e.g. when it
+		// was changed, accidentally or not). Also good for audit trail.
+		saveGenesisDoc(stateDB, genDoc)
+	}
+
+	state, err := sm.LoadStateFromDBOrGenesisDoc(stateDB, genDoc)
+	if err != nil {
+		return err
+	}
+
+	if config.RollbackFlag {
+		unsafeRollbackData(config, state, stateDB, blockStore, blockStoreDB, privValidator)
+	}
+
+	return nil
+}
+
+func loadGenesisDoc(db dbm.DB) (*types.GenesisDoc, error) {
+	bytes := db.Get(genesisDocKey)
+	if len(bytes) == 0 {
+		return nil, cmn.ErrorWrap(nil, "Genesis doc not found")
+	}
+	var genDoc *types.GenesisDoc
+	err := cdc.UnmarshalJSON(bytes, &genDoc)
+	if err != nil {
+		return nil, cmn.ErrorWrap(err, "Failed to load genesis doc due to unmarshaling error: %v (bytes: %X)", err, bytes)
+	}
+	return genDoc, nil
+}
+
+// panics if failed to marshal the given genesis document
+func saveGenesisDoc(db dbm.DB, genDoc *types.GenesisDoc) {
+	bytes, err := cdc.MarshalJSON(genDoc)
+	if err != nil {
+		panic("Failed to load genesis doc due to unmarshaling error")
+	}
+	db.SetSync(genesisDocKey, bytes)
+}
+
+func unsafeRollbackData(config *cfg.Config, state sm.State, stateDb dbm.DB, blockStore *bc.BlockStore,
+	blockStoreDB dbm.DB, privValidator types.PrivValidator) error {
+
+	if state.LastBlockHeight > config.BaseConfig.RollbackHeight {
+		stateRestore := restoreStateFromBlock(stateDb, blockStore, config.BaseConfig.RollbackHeight)
+		sm.SaveState(stateDb, stateRestore)
+		bc.BlockStoreStateJSON{Height: config.BaseConfig.RollbackHeight}.Save(blockStoreDB)
+		modifyPrivValidatorsFile(config, config.BaseConfig.RollbackHeight)
+		blockStore = bc.NewBlockStore(blockStoreDB)
+		return nil
+	} else {
+	}
+	return nil
+}
+
+func restoreStateFromBlock(stateDb dbm.DB, blockStore *bc.BlockStore, rollbackHeight int64) sm.State {
+	validator, _ := sm.LoadValidators(stateDb, rollbackHeight+1)
+	validatorChanged := sm.LoadValidatorsChanged(stateDb, rollbackHeight)
+	lastvalidator, _ := sm.LoadValidators(stateDb, rollbackHeight)
+	nextvalidator, _ := sm.LoadValidators(stateDb, rollbackHeight+2)
+
+	consensusParams, _ := sm.LoadConsensusParams(stateDb, rollbackHeight)
+	consensusParamsChanged := sm.LoadConsensusParamsChanged(stateDb, rollbackHeight)
+	software := sm.LoadSoftware(stateDb, rollbackHeight)
+
+	block := blockStore.LoadBlock(rollbackHeight)
+	nextBlock := blockStore.LoadBlock(rollbackHeight + 1)
+
+	return sm.State{
+		ChainID: block.ChainID,
+		Version: sm.Version{Consensus: block.Version, Software: software},
+
+		LastBlockID:      nextBlock.LastBlockID, //? true
+		LastBlockHeight:  block.Height,
+		LastBlockTime:    block.Time,
```

Comment:
LoadValidators can't get correct proposer priority and proposer, releted to https://github.com/tendermint/tendermint/pull/3438.
If a validator rollback, the consensus will in chaos.

Reply:
I saved it every block from 
```
	if lastHeightChanged == height {
		valInfo.ValidatorSet = valSet
	}	
```
to
```
valInfo.ValidatorSet = valSet
```
whether lastHeightChanged is true or false.

Reply:
I can't say whether  it's bad or right. For it will change the storage of statedb.
And may cause some performance.